search.py

```diff
@@ -406,101 +406,92 @@ def astar_search(problem, h=None):
-        in any given state of the environment."""
-       
-        possible_actions = list()
+        in any given state of the environment '''
+        
+        possible_actions = ['UP', 'DOWN', 'LEFT', 'RIGHT']       
         index_blank_square = self.find_blank_square(state)
 
-        if index_blank_square(0) == 0:
-            possible_actions += ['DOWN']
-        elif index_blank_square(0) == 1:
-            possible_actions += ['UP', 'DOWN']
-        elif index_blank_square(0) == 2:
-            possible_actions += ['UP']
-        
-        if index_blank_square(1) == 0:
-            possible_actions += ['RIGHT']
-        elif index_blank_square(1) == 1:
-            possible_actions += ['LEFT', 'RIGHT']
-        elif index_blank_square(1) == 2:
-            possible_actions += ['LEFT']
+        if index_blank_square % 3 == 0:
+            possible_actions.remove('LEFT')
+        if index_blank_square < 3:
+            possible_actions.remove('UP')
+        if index_blank_square % 3 == 2:
+            possible_actions.remove('RIGHT')
+        if index_blank_square > 5:
+            possible_actions.remove('DOWN')
 
         return possible_actions
 
     def result(self, state, action):
-        """Given state and action, return a new state that is the result of the action.
-        Action is assumed to be a valid action in the state."""
-
-        blank_square = self.find_blank_square(state)
-        new_state = [row[:] for row in state]
-
-        if action=='UP':
-            new_state[blank_square(0)][blank_square(1)] = new_state[blank_square(0)-1][blank_square(1)]
-            new_state[blank_square(0)-1][blank_square(1)] = 0
-        elif action=='LEFT':
-            new_state[blank_square(0)][blank_square(1)] = new_state[blank_square(0)][blank_square(1)-1]
-            new_state[blank_square(0)][blank_square(1)-1] = 0
-        elif action=='DOWN':
-            new_state[blank_square(0)][blank_square(1)] = new_state[blank_square(0)+1][blank_square(1)]
-            new_state[blank_square(0)+1][blank_square(1)] = 0
-        elif action=='RIGHT':
-            new_state[blank_square(0)][blank_square(1)] = new_state[blank_square(0)][blank_square(1)+1]
-            new_state[blank_square(0)][blank_square(1)+1] = 0
+        ''' Given state and action, return a new state that is the result of the action.
+        Action is assumed to be a valid action in the state '''
+
+        # ix is the index of the blank square
+        ix = self.find_blank_square(state)
+        new_state = list(state)
+
+        if action == 'UP':
+            new_state[ix], new_state[ix - 3] = new_state[ix - 3], 0
+        elif action == 'DOWN':
+            new_state[ix], new_state[ix + 3] = new_state[ix + 3], 0
+        elif action == 'LEFT':
+            new_state[ix], new_state[ix - 1] = new_state[ix - 1], 0
+        elif action == 'RIGHT':
+            new_state[ix], new_state[ix + 1] = new_state[ix + 1], 0
         else:
```

Comment:
I would say just ignore an illegal action, but if you are going to print something, say what the invalid action is.